sfs/writer: Error handling and recovery

irq0 opened the pull request:
Depends on  #124
Depends on  #135 

Replace iostream code with cstyle alternatives that don't throw and are closer to how ceph src/os does IO. Handle errors during open/close/write by deleting the failed file and returning an error.

Convert errno errors to S3 errors where it makes sense.

fsync directories after creating deleting the object file.

https://github.com/aquarist-labs/s3gw/issues/363

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

Commit:
rgw/sfs: atomic writer error handling and recovery

Replace iostream code with cstyle alternatives that don't throw and
are closer to how ceph src/os does IO. Handle errors during
open/close/write by deleting the failed file and returning an error.

Convert errno errors to S3 errors where it makes sense.

fsync directories after creating deleting the object file.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

irq0 commented:
Looking for some general feedback and there is one thing open:
What to do with the database metadata on failure?
Right now the patch deletes the data, but keeps the metadata in state WRITING

1. keep it - let the GC remove partial objects. mark "FAILED"
2. remove on failure
3. only create after the write is finished

I tend to favor 1., because we could easily display a list of failed uploads and their reasons on a status page

jecluis commented on src/rgw/driver/sfs/writer.cc:
```diff
@@ -38,13 +43,109 @@ SFSAtomicWriter::SFSAtomicWriter(
+  int result = 0;
+  int ret;
+
+  ret = ::fsync(fd);
```
maybe assert before that `fd >= 0`?

irq0 replied:
:+1:

jecluis commented on src/rgw/driver/sfs/writer.cc:
```diff
@@ -38,13 +43,109 @@ SFSAtomicWriter::SFSAtomicWriter(
+  const auto dir_fd = ::open(object_path.parent_path().c_str(), O_RDONLY);
+  int ret = ::fsync(dir_fd);
+  if (ret < 0) {
+    lsfs_dout(dpp, 10)
```
if we're logging failures, might be better to log them to level `-1`. Would be nice to have defined an `lsfs_derr()` macro, but given that's not available, logging to `-1` would suffice. Otherwise these errors will be hidden most of the time in a production environment.

irq0 replied:
:+1:

jecluis commented on src/rgw/driver/sfs/writer.cc:
```diff
@@ -59,48 +160,86 @@ int SFSAtomicWriter::prepare(optional_yield y) {
-  ceph_assert(std::filesystem::exists(object_path));
-
-  lsfs_dout(dpp, 10) << "write to object at " << object_path << dendl;
+  lsfs_dout(dpp, 10) << "data len: " << data.length() << ", offset: " << offset
```
why are we using `fmt::format()` pretty much everywhere else, but falling back to the `<<` operator here?

irq0 replied:
no reason. Adding a commit that converts the longer log lines to fmt

jecluis reviewed (changes requested):
Generally looks good, but I think the errors need to be logged at `-1` instead of `10` so they are not lost when running in production environments.

Also needs tests.

jecluis commented:
> Looking for some general feedback and there is one thing open: What to do with the database metadata on failure? Right now the patch deletes the data, but keeps the metadata in state WRITING
> 
> 1. keep it - let the GC remove partial objects. mark "FAILED"
> 2. remove on failure
> 3. only create after the write is finished
> 
> I tend to favor 1., because we could easily display a list of failed uploads and their reasons on a status page

I would agree, although in case of an `ENOSPC` that might not be feasible. I'm not sure how we're going to update the DB if an `ENOSPC` is hit. Hopefully we stop writing once the disk is at `ENOSPC - some%` to avoid a situation where we're no longer able to write/rm stuff, but from the looks of this patch set it seems to me that if we hit `ENOSPC`, then it's because the underlying system call is throwing `ENOSPC` for reals.

Commit:
rgw/sfs: Remove object state WRITING

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rgw/sfs: Remove state changes to WRITING

State OPEN from which we quickly transitioned to WRITING takes its
role. This saves us from state change calls in the hot write paths.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

irq0 commented:
I'm putting this on hold until #124 is in

jecluis commented:
> I'm putting this on hold until #124 is in

making it a draft until then.

Commit:
rgw/sfs: Log atomic writer errors with level -1

Increases likelihood that errors reach logs/user in prod environments.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rgw/sfs: Use fmt in longer atomic reader log msg

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rgw/sfs: clean up in flight version on atomic writer failure

Remove version created for the atomic writer when an error forced the
writer into the cleanup path.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

irq0 commented:
> Also needs tests.

Unit testing would be painful, as we would need to mock / fake the io functions, no?

We probably get most bang for buck with adding a test suite that uses an error injection filesystem. We need to do this eventually to at minimum ensure that we don't crash under IO errors, etc.

Charybdefs has a nice example that fits the ENOSPC our use case that illustrates what I mean: https://github.com/scylladb/charybdefs/blob/master/tests/out_of_disk_space_test.py

jecluis commented:
I'll be happy to re-review this once the other 2 PRs are merged.

jhmarina commented:
All the PRs mentioned have been merged. @jecluis @irq0 is this ready for a re-review?

irq0 commented:
> All the PRs mentioned have been merged. @jecluis @irq0 is this ready for a re-review?

Yes

jecluis reviewed (approved):
lgtm but with a caveat: IMO we should cleanup the written file only when we are the ones creating it. If the file already existed (and assuming no versioning enabled), we should not be removing the file during cleanup.
